File: src/tools/semgrep/eval-repos/synthetic/python/performance_issues.py
"""
Test file for PERFORMANCE_ISSUE detection.
Tests string concat in loop, list += in loop, regex in loop.
"""

import re
from typing import List


# 1. String concatenation in loop
def build_string_inefficient(items):
    """PERFORMANCE: String concat in loop creates many intermediate strings."""
    result = ""
    for item in items:
        result += str(item) + ","  # O(n^2) performance
    return result


# 2. List extend with += in loop (less efficient)
def extend_list_inefficient(lists):
    """PERFORMANCE: Using += to extend list in loop is inefficient."""
    result = []
    for lst in lists:
        result += lst  # Creates new list each iteration
    return result


# 3. Regex compilation in loop
def search_patterns_inefficient(texts, pattern):
    """PERFORMANCE: Compiling regex in loop is wasteful."""
    matches = []
    for text in texts:
        if re.search(pattern, text):  # Regex compiled each iteration
            matches.append(text)
    return matches


# 4. List comprehension with repeated function call
def process_items_inefficient(items):
    """PERFORMANCE: Repeated len() calls in loop condition."""
    i = 0
    while i < len(items):  # len() called each iteration
        process(items[i])
        i += 1


# 5. Using + for list concatenation in loop
def merge_lists_inefficient(lists):
    """PERFORMANCE: List + in loop creates many intermediate lists."""
    result = []
    for lst in lists:
        result = result + lst  # O(n^2) performance
    return result


# GOOD: Correct string building
def build_string_correct(items):
    """Good: Use join() for string building."""
    return ",".join(str(item) for item in items)


# GOOD: Correct list extension
def extend_list_correct(lists):
    """Good: Use list.extend() or itertools.chain."""
    result = []
    for lst in lists:
        result.extend(lst)
    return result


# GOOD: Pre-compiled regex
PATTERN = re.compile(r'\d+')

def search_patterns_correct(texts):
    """Good: Compile regex once, use many times."""
    matches = []
    for text in texts:
        if PATTERN.search(text):
            matches.append(text)
    return matches


# GOOD: Cache length outside loop
def process_items_correct(items):
    """Good: Cache len() result outside loop."""
    length = len(items)
    i = 0
    while i < length:
        process(items[i])
        i += 1


def process(item):
    """Stub function."""
    pass

File: src/tools/semgrep/eval-repos/synthetic/csharp/PerformanceIssues.cs
/**
 * Test file for PERFORMANCE_ISSUE detection.
 * Tests string concat loop, Count() vs Any(), uncached regex.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace SmellTests.Performance
{
    public class PerformanceIssues
    {
        // 1. String concatenation in loop
        public string BuildStringInefficient(IEnumerable<string> items)
        {
            // PERFORMANCE: String concat in loop creates many intermediate strings
            string result = "";
            foreach (var item in items)
            {
                result += item + ",";
            }
            return result;
        }

        // 2. Using Count() when Any() would suffice
        public bool HasItemsInefficient(IEnumerable<int> items)
        {
            // PERFORMANCE: Count() iterates entire collection
            return items.Count() > 0;
        }

        // 3. Regex not cached
        public bool MatchesPatternInefficient(string input)
        {
            // PERFORMANCE: Regex.IsMatch compiles pattern each call
            return Regex.IsMatch(input, @"\d{4}-\d{2}-\d{2}");
        }

        // 4. LINQ query in loop
        public List<int> FilterInLoop(List<int> items, int[] thresholds)
        {
            var results = new List<int>();
            // PERFORMANCE: LINQ query executed multiple times in loop
            foreach (var threshold in thresholds)
            {
                results.AddRange(items.Where(x => x > threshold));
            }
            return results;
        }

        // 5. Boxing in loop
        public void ProcessNumbers(int[] numbers)
        {
            // PERFORMANCE: Boxing integers in loop
            var list = new List<object>();
            foreach (var n in numbers)
            {
                list.Add(n);  // Boxing occurs here
            }
        }

        // GOOD: StringBuilder for string building
        public string BuildStringCorrect(IEnumerable<string> items)
        {
            var sb = new StringBuilder();
            foreach (var item in items)
            {
                sb.Append(item).Append(',');
            }
            return sb.ToString();
        }

        // GOOD: Use Any() for existence check
        public bool HasItemsCorrect(IEnumerable<int> items)
        {
            return items.Any();
        }

        // GOOD: Cached regex
        private static readonly Regex DatePattern = new Regex(
            @"\d{4}-\d{2}-\d{2}",
            RegexOptions.Compiled);

        public bool MatchesPatternCorrect(string input)
        {
            return DatePattern.IsMatch(input);
        }

        // GOOD: Generic list to avoid boxing
        public void ProcessNumbersCorrect(int[] numbers)
        {
            var list = new List<int>();
            foreach (var n in numbers)
            {
                list.Add(n);  // No boxing
            }
        }
    }
}
